File: python3/linked_list_random_node.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import random
class Solution:
    def __init__(self, head: ListNode):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        """
        self.mem = []
        
        while head:
            self.mem.append(head.val)
            head = head.next

    def getRandom(self) -> int:
        """
        Returns a random node's value.
        """
        idx = random.randint(0, len(self.mem)-1)
        return self.mem[idx]
        
        
# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()

File: golang/linked_list_random_node.go
package main

// Space	: O(1)
// Time	: O(n)

import (
	"golang/model"
	"math/rand"
)

type Solution struct {
	LinkedList *model.ListNode
	Length     int
}

func Constructor(head *model.ListNode) Solution {
	p := head
	count := 0
	for p != nil {
		count += 1
		p = p.Next
	}

	return Solution{
		LinkedList: head,
		Length:     count,
	}
}

func (this *Solution) GetRandom() int {
	r := rand.Intn(this.Length)
	p := this.LinkedList
	for r > 0 {
		r--
		p = p.Next
	}
	return p.Val
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(head);
 * param_1 := obj.GetRandom();
 */
